File: webhook-http-url/python/github_hook_transport.py
"""Find webhooks that deliver over plaintext HTTP, and say which ones leak.

Read only. Every call is a GET. Changing a hook's URL is a write and this
script does not do it: it prints the change, as a full config rather than a
single field, because a webhook's config is replaced rather than merged and the
secret you read back is a mask.

GitHub will deliver to an http:// URL without complaint. The payload and the
signature header that authenticates it both cross the network as plain text. A
signature proves the payload came from GitHub; it conceals nothing and it can
be replayed by anyone who saw it.

Two things that look the same in this field are kept apart:

    http:// on a routable host      payloads are readable in transit
    http:// on a private address    GitHub cannot route there at all, so this
                                    hook has never delivered anything

And one thing that hides here is named: insecure_ssl reads "0" on a plaintext
hook, because there is no certificate to verify when there is no TLS, so the
field an audit samples reports the reassuring value.

The secret is never printed, and neither is any query string or userinfo in a
hook URL.

Environment:

    GITHUB_TOKEN    a read-only token that can see the repository's hooks
"""
import argparse
import json
import logging
import os
import sys

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("github_hook_transport")

API = "https://api.github.com"
UA = "github-hook-transport/1.0"

# Names that never resolve to somewhere GitHub can deliver.
LOCAL_NAMES = ("localhost", "localhost.localdomain", "ip6-localhost")
LOCAL_SUFFIXES = (".localhost", ".local", ".internal", ".lan", ".home.arpa",
                  ".localdomain")

# States that mean payloads are readable in transit.
LEAKING = ("plaintext",)


def config_of(hook):
    """The config object of a hook, or an empty dict. Pure."""
    if not isinstance(hook, dict):
        return {}
    config = hook.get("config")
    return config if isinstance(config, dict) else {}


def raw_url(hook):
    """The configured URL, trimmed, or "". Pure."""
    return str(config_of(hook).get("url") or "").strip()


def safe_url(url):
    """A URL with its query string and any userinfo removed. Pure.

    This script prints URLs, and those are the two places a credential hides in
    one: a token in a query parameter and a user:pass pair before the host.
    """
    text = str(url or "").strip()
    if not text:
        return ""
    text = text.split("?", 1)[0].split("#", 1)[0]
    if "://" in text:
        scheme, rest = text.split("://", 1)
    else:
        scheme, rest = "", text
    if "@" in rest:
        rest = rest.rsplit("@", 1)[1]
        rest = "<redacted>@" + rest
    return ("%s://%s" % (scheme, rest)) if scheme else rest


def scheme_of(url):
    """The lower-cased scheme of a URL, or "". Pure."""
    text = str(url or "").strip()
    return text.split("://", 1)[0].lower() if "://" in text else ""


def host_of(url):
    """The lower-cased host of a URL, without port or userinfo. Pure."""
    text = str(url or "").strip()
    rest = text.split("://", 1)[1] if "://" in text else text
    rest = rest.split("/", 1)[0]
    if "@" in rest:
        rest = rest.rsplit("@", 1)[1]
    if rest.startswith("["):
        return rest[1:].split("]", 1)[0].lower()
    return rest.split(":", 1)[0].lower()


def is_private_host(host):
    """Whether a host is somewhere GitHub's delivery network cannot reach. Pure.

    Deliberately a name and address test rather than a DNS lookup. A resolver
    inside your network answers differently from GitHub's, so resolving here
    would produce an answer about the wrong network, and a script that makes
    DNS queries for every hook in an organization is a different kind of tool.
    """
    name = str(host or "").strip().lower().strip(".")
    if not name:
        return False
    if name in LOCAL_NAMES or name.endswith(LOCAL_SUFFIXES):
        return True
    if name in ("::1", "0:0:0:0:0:0:0:1"):
        return True
    if name.startswith(("fd", "fc", "fe80:")) and ":" in name:
        return True
    parts = name.split(".")
    if len(parts) == 4 and all(p.isdigit() and len(p) <= 3 for p in parts):
        octets = [int(p) for p in parts]
        if any(o > 255 for o in octets):
            return False
        if octets[0] in (0, 127, 10):
            return True
        if octets[0] == 192 and octets[1] == 168:
            return True
        if octets[0] == 172 and 16 <= octets[1] <= 31:
            return True
        if octets[0] == 169 and octets[1] == 254:
            return True
    return False


def insecure_ssl_reads(hook):
    """The insecure_ssl value as text, or "" when absent. Pure.

    Read as text on purpose. This script does not interpret the flag - that is
    the certificate-verification question - it only needs to report what an
    audit sampling the field would have seen.
    """
    config = config_of(hook)
    if "insecure_ssl" not in config:
        return ""
    return str(config["insecure_ssl"]).strip().lower()


def looks_compliant(hook):
    """Whether a plaintext hook reads as safe on the field audits sample. Pure.

    The whole reason this problem survives a review that was genuinely carried
    out: insecure_ssl is "0" on a hook with no TLS, because there is no
    certificate to verify.
    """
    return (scheme_of(raw_url(hook)) not in ("https", "")
            and insecure_ssl_reads(hook) in ("0", "false"))


def has_secret(hook):
    """Whether the hook has a secret set. Pure. The value is never read."""
    return "secret" in config_of(hook)


def classify(hook):
    """Sort one hook into a state and a sentence. Pure."""
    ident = "hook %s" % (hook.get("id", "?") if isinstance(hook, dict) else "?")
    url = raw_url(hook)
    scheme = scheme_of(url)
    if not url or not scheme:
        return ("no-scheme",
                "%s has no usable URL in its config, so nothing can be said "
                "about how it delivers." % ident)
    if scheme == "https":
        if insecure_ssl_reads(hook) in ("1", "true"):
            return ("encrypted-unverified",
                    "%s posts to %s over TLS, but with certificate "
                    "verification disabled. The transport is encrypted and "
                    "unauthenticated, which is a different question from this "
                    "one." % (ident, safe_url(url)))
        return ("encrypted",
                "%s posts to %s over TLS." % (ident, safe_url(url)))
    if scheme != "http":
        return ("unknown-scheme",
                "%s posts to a %s:// URL, which is not a scheme GitHub "
                "delivers to. Read the URL by hand." % (ident, scheme))
    if is_private_host(host_of(url)):
        return ("plaintext-unreachable",
                "%s posts to %s, which GitHub cannot route to. This hook has "
                "never delivered anything, and it is not leaking payloads "
                "either." % (ident, safe_url(url)))
    suffix = (" insecure_ssl reads \"%s\", which is what a hook with no TLS "
              "at all always reads." % insecure_ssl_reads(hook)
              if looks_compliant(hook) else "")
    return ("plaintext",
            "%s posts to %s over an unencrypted connection.%s"
            % (ident, safe_url(url), suffix))


def repair(state, hook):
    """The change to make, printed as a whole config. Pure."""
    if state == "plaintext":
        rotate = (" and a new secret. Rotate: that secret has been signing "
                  "payloads on an open channel." if has_secret(hook)
                  else " and a secret, since this hook has none.")
        return ("move the receiver behind HTTPS, then send the hook's full "
                "config with the new URL, the content type%s The config is "
                "replaced, not merged, and the secret you read back is a mask."
                % rotate)
    if state == "plaintext-unreachable":
        return ("delete this hook, or point it at an endpoint GitHub can "
                "reach over HTTPS. Its delivery log will be connection errors "
                "and timeouts for as far back as the retention window goes.")
    if state == "encrypted-unverified":
        return ("this is the certificate-verification question rather than "
                "the transport one. Fix the certificate, then set insecure_ssl "
                "back to \"0\" as part of a full config update.")
    if state in ("no-scheme", "unknown-scheme"):
        return ("read the hook's URL by hand. A hook GitHub cannot parse a "
                "scheme from is not delivering anything.")
    return "nothing. This hook delivers over TLS."


def summarize(hooks):
    """Counts across every hook read. Pure."""
    rows = [h for h in (hooks or []) if isinstance(h, dict)]
    states = [classify(h)[0] for h in rows]
    return {"total": len(rows),
            "plaintext": states.count("plaintext"),
            "unreachable": states.count("plaintext-unreachable"),
            "encrypted": states.count("encrypted") + states.count("encrypted-unverified"),
            "unreadable": states.count("no-scheme") + states.count("unknown-scheme")}


def get(session, path):
    """One GET. Returns (status, json-or-None)."""
    url = API + path if path.startswith("/") else path
    r = session.get(url, timeout=30)
    try:
        body = r.json()
    except ValueError:
        body = None
    return r.status_code, body


def list_hooks(session, scope):
    """Hooks for a repo (owner/name) or an org (@org). Read only."""
    path = ("/orgs/%s/hooks?per_page=100" % scope[1:] if scope.startswith("@")
            else "/repos/%s/hooks?per_page=100" % scope)
    status, body = get(session, path)
    if status != 200 or not isinstance(body, list):
        log.error("GET %s returned %d", path, status)
        return []
    return body


def main():
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--repo", action="append", default=[],
                    help="owner/name; repeatable")
    ap.add_argument("--org", action="append", default=[],
                    help="organization login; repeatable")
    args = ap.parse_args()

    token = os.environ.get("GITHUB_TOKEN")
    if not token:
        log.error("set GITHUB_TOKEN to a read-only token that can see the "
                  "repository's hooks")
        return 2
    scopes = list(args.repo) + ["@" + o for o in args.org]
    if not scopes:
        log.error("pass at least one --repo owner/name or --org login")
        return 2

    session = requests.Session()
    session.headers.update({
        "Authorization": "Bearer " + token,
        "Accept": "application/vnd.github+json",
        "X-GitHub-Api-Version": "2022-11-28",
        "User-Agent": UA,
    })

    findings = []
    for scope in scopes:
        label = scope[1:] if scope.startswith("@") else scope
        hooks = list_hooks(session, scope)
        stats = summarize(hooks)
        log.info("%d hook(s) on %s", stats["total"], label)
        for hook in hooks:
            state, detail = classify(hook)
            findings.append({"scope": label, "hook_id": hook.get("id"),
                             "state": state, "detail": detail,
                             "url": safe_url(raw_url(hook)),
                             "looks_compliant": looks_compliant(hook)})
            if state != "encrypted":
                log.info("%s: %s", state, detail)
                log.info("repair: %s", repair(state, hook))
        if stats["plaintext"] == 0:
            log.info("encrypted: no hook on %s delivers over plaintext HTTP "
                     "to a routable host", label)

    print(json.dumps({"scopes": scopes, "findings": findings},
                     indent=2, default=str))
    return 1 if any(f["state"] in LEAKING for f in findings) else 0


if __name__ == "__main__":
    sys.exit(main())

File: webhook-http-url/node/github-hook-transport.mjs
/**
 * Find webhooks that deliver over plaintext HTTP, and say which ones leak.
 *
 * Read only. Every call is a GET. Changing a hook's URL is a write and is not
 * done here: the script prints the change, as a full config rather than a
 * single field, because a webhook's config is replaced rather than merged and
 * the secret you read back is a mask.
 *
 * Two things that look the same in this field are kept apart: http:// on a
 * routable host means payloads are readable in transit, while http:// on a
 * private address means GitHub cannot route there at all and the hook has
 * never delivered anything.
 *
 * The secret is never printed, and neither is any query string or userinfo in
 * a hook URL.
 *
 * Environment:
 *   GITHUB_TOKEN   a read-only token that can see the repository's hooks
 *
 * Usage:
 *   node github-hook-transport.mjs acme-corp/api @acme-corp
 */
const API = 'https://api.github.com';
const UA = 'github-hook-transport/1.0';

const LOCAL_NAMES = ['localhost', 'localhost.localdomain', 'ip6-localhost'];
const LOCAL_SUFFIXES = ['.localhost', '.local', '.internal', '.lan',
  '.home.arpa', '.localdomain'];

/** States that mean payloads are readable in transit. */
export const LEAKING = ['plaintext'];

/** The config object of a hook, or an empty object. Pure. */
export function configOf(hook) {
  if (!hook || typeof hook !== 'object') return {};
  const config = hook.config;
  return config && typeof config === 'object' ? config : {};
}

/** The configured URL, trimmed, or ''. Pure. */
export function rawUrl(hook) {
  return String(configOf(hook).url ?? '').trim();
}

/** A URL with its query string and any userinfo removed. Pure. */
export function safeUrl(url) {
  let text = String(url ?? '').trim();
  if (!text) return '';
  text = text.split('?')[0].split('#')[0];
  let scheme = '';
  let rest = text;
  if (text.includes('://')) {
    const idx = text.indexOf('://');
    scheme = text.slice(0, idx);
    rest = text.slice(idx + 3);
  }
  if (rest.includes('@')) {
    rest = `<redacted>@${rest.slice(rest.lastIndexOf('@') + 1)}`;
  }
  return scheme ? `${scheme}://${rest}` : rest;
}

/** The lower-cased scheme of a URL, or ''. Pure. */
export function schemeOf(url) {
  const text = String(url ?? '').trim();
  return text.includes('://') ? text.slice(0, text.indexOf('://')).toLowerCase() : '';
}

/** The lower-cased host of a URL, without port or userinfo. Pure. */
export function hostOf(url) {
  const text = String(url ?? '').trim();
  let rest = text.includes('://') ? text.slice(text.indexOf('://') + 3) : text;
  rest = rest.split('/')[0];
  if (rest.includes('@')) rest = rest.slice(rest.lastIndexOf('@') + 1);
  if (rest.startsWith('[')) return rest.slice(1).split(']')[0].toLowerCase();
  return rest.split(':')[0].toLowerCase();
}

/**
 * Whether a host is somewhere GitHub's delivery network cannot reach. Pure.
 * A name and address test rather than a DNS lookup: a resolver inside your
 * network answers differently from GitHub's.
 */
export function isPrivateHost(host) {
  const name = String(host ?? '').trim().toLowerCase().replace(/^\.+|\.+$/g, '');
  if (!name) return false;
  if (LOCAL_NAMES.includes(name)) return true;
  if (LOCAL_SUFFIXES.some((s) => name.endsWith(s))) return true;
  if (name === '::1' || name === '0:0:0:0:0:0:0:1') return true;
  if (name.includes(':') && (name.startsWith('fd') || name.startsWith('fc') || name.startsWith('fe80:'))) {
    return true;
  }
  const parts = name.split('.');
  if (parts.length === 4 && parts.every((p) => /^[0-9]{1,3}$/.test(p))) {
    const o = parts.map(Number);
    if (o.some((n) => n > 255)) return false;
    if ([0, 127, 10].includes(o[0])) return true;
    if (o[0] === 192 && o[1] === 168) return true;
    if (o[0] === 172 && o[1] >= 16 && o[1] <= 31) return true;
    if (o[0] === 169 && o[1] === 254) return true;
  }
  return false;
}

/** The insecure_ssl value as text, or '' when absent. Pure. */
export function insecureSslReads(hook) {
  const config = configOf(hook);
  if (!('insecure_ssl' in config)) return '';
  return String(config.insecure_ssl).trim().toLowerCase();
}

/** Whether a plaintext hook reads as safe on the field audits sample. Pure. */
export function looksCompliant(hook) {
  const scheme = schemeOf(rawUrl(hook));
  return scheme !== 'https' && scheme !== '' && ['0', 'false'].includes(insecureSslReads(hook));
}

/** Whether the hook has a secret set. Pure. The value is never read. */
export function hasSecret(hook) {
  return 'secret' in configOf(hook);
}

/** Sort one hook into a state and a sentence. Pure. */
export function classify(hook) {
  const ident = `hook ${(hook && typeof hook === 'object' ? hook.id : null) ?? '?'}`;
  const url = rawUrl(hook);
  const scheme = schemeOf(url);
  if (!url || !scheme) {
    return ['no-scheme',
      `${ident} has no usable URL in its config, so nothing can be said about ` +
      'how it delivers.'];
  }
  if (scheme === 'https') {
    if (['1', 'true'].includes(insecureSslReads(hook))) {
      return ['encrypted-unverified',
        `${ident} posts to ${safeUrl(url)} over TLS, but with certificate ` +
        'verification disabled. The transport is encrypted and ' +
        'unauthenticated, which is a different question from this one.'];
    }
    return ['encrypted', `${ident} posts to ${safeUrl(url)} over TLS.`];
  }
  if (scheme !== 'http') {
    return ['unknown-scheme',
      `${ident} posts to a ${scheme}:// URL, which is not a scheme GitHub ` +
      'delivers to. Read the URL by hand.'];
  }
  if (isPrivateHost(hostOf(url))) {
    return ['plaintext-unreachable',
      `${ident} posts to ${safeUrl(url)}, which GitHub cannot route to. This ` +
      'hook has never delivered anything, and it is not leaking payloads either.'];
  }
  const suffix = looksCompliant(hook)
    ? ` insecure_ssl reads "${insecureSslReads(hook)}", which is what a hook ` +
      'with no TLS at all always reads.'
    : '';
  return ['plaintext',
    `${ident} posts to ${safeUrl(url)} over an unencrypted connection.${suffix}`];
}

/** The change to make, printed as a whole config. Pure. */
export function repair(state, hook) {
  if (state === 'plaintext') {
    const rotate = hasSecret(hook)
      ? ' and a new secret. Rotate: that secret has been signing payloads on an open channel.'
      : ' and a secret, since this hook has none.';
    return 'move the receiver behind HTTPS, then send the hook\'s full config ' +
      `with the new URL, the content type${rotate} The config is replaced, ` +
      'not merged, and the secret you read back is a mask.';
  }
  if (state === 'plaintext-unreachable') {
    return 'delete this hook, or point it at an endpoint GitHub can reach ' +
      'over HTTPS. Its delivery log will be connection errors and timeouts ' +
      'for as far back as the retention window goes.';
  }
  if (state === 'encrypted-unverified') {
    return 'this is the certificate-verification question rather than the ' +
      'transport one. Fix the certificate, then set insecure_ssl back to "0" ' +
      'as part of a full config update.';
  }
  if (state === 'no-scheme' || state === 'unknown-scheme') {
    return "read the hook's URL by hand. A hook GitHub cannot parse a scheme " +
      'from is not delivering anything.';
  }
  return 'nothing. This hook delivers over TLS.';
}

/** Counts across every hook read. Pure. */
export function summarize(hooks) {
  const rows = (hooks ?? []).filter((h) => h && typeof h === 'object');
  const states = rows.map((h) => classify(h)[0]);
  const count = (name) => states.filter((s) => s === name).length;
  return {
    total: rows.length,
    plaintext: count('plaintext'),
    unreachable: count('plaintext-unreachable'),
    encrypted: count('encrypted') + count('encrypted-unverified'),
    unreadable: count('no-scheme') + count('unknown-scheme'),
  };
}

async function get(token, path) {
  const res = await fetch(path.startsWith('/') ? API + path : path, {
    headers: {
      Authorization: `Bearer ${token}`,
      Accept: 'application/vnd.github+json',
      'X-GitHub-Api-Version': '2022-11-28',
      'User-Agent': UA,
    },
  });
  let body = null;
  try { body = await res.json(); } catch { body = null; }
  return { status: res.status, body };
}

async function listHooks(token, scope) {
  const path = scope.startsWith('@')
    ? `/orgs/${scope.slice(1)}/hooks?per_page=100`
    : `/repos/${scope}/hooks?per_page=100`;
  const { status, body } = await get(token, path);
  if (status !== 200 || !Array.isArray(body)) {
    console.error(`GET ${path} returned ${status}`);
    return [];
  }
  return body;
}

async function main() {
  const token = (process.env.GITHUB_TOKEN || "dummy-github-token");
  if (!token) {
    console.error('set GITHUB_TOKEN to a read-only token that can see the ' +
      "repository's hooks");
    process.exitCode = 2;
    return;
  }
  const scopes = process.argv.slice(2);
  if (scopes.length === 0) {
    console.error('pass at least one owner/name, or @org for an organization');
    process.exitCode = 2;
    return;
  }

  const findings = [];
  for (const scope of scopes) {
    const label = scope.startsWith('@') ? scope.slice(1) : scope;
    const hooks = await listHooks(token, scope);
    const stats = summarize(hooks);
    console.log(`${stats.total} hook(s) on ${label}`);
    for (const hook of hooks) {
      const [state, detail] = classify(hook);
      findings.push({
        scope: label,
        hook_id: hook.id,
        state,
        detail,
        url: safeUrl(rawUrl(hook)),
        looks_compliant: looksCompliant(hook),
      });
      if (state !== 'encrypted') {
        console.log(`${state}: ${detail}`);
        console.log(`repair: ${repair(state, hook)}`);
      }
    }
    if (stats.plaintext === 0) {
      console.log(`encrypted: no hook on ${label} delivers over plaintext ` +
        'HTTP to a routable host');
    }
  }

  console.log(JSON.stringify({ scopes, findings }, null, 2));
  process.exitCode = findings.some((f) => LEAKING.includes(f.state)) ? 1 : 0;
}

// Only run when invoked directly, so importing this module from the test file
// does not execute main() and set an exit code that fails the suite even as
// every test passes.
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((err) => { console.error(err.message); process.exitCode = 2; });
}
